_posts/2017/2017-16-01-testing-async-interactions.md

```diff
@@ -0,0 +1,246 @@
+We call such a test a local integration test.
+
+A test that is fairly easy to implement on checkout-service side is to just hard-code the event payload as the checkout service expects it and assert that the service correctly consumes this event.
+
+Such a [test](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/test/java/com/epages/checkout/ProductSubscriberHardcodedPayloadTest.java) should do two things:
+
+- make sure that the JSON event payload can be deserialized correctly
+- assert that the service is reacting appropriately
+
+```java
+@Test
+public void should_handle_product_created_event() {
+  //GIVEN
+  String payload = "{\n" +
+    "  \"id\": 8,\n" +
+    "  \"name\": \"Awesome Jeans\",\n" +
+    "  \"salesPrice\": 49.99\n" +
+    "}";
+
+  //WHEN
+  createAndEmitEvent(payload);
+  //THEN
+  then(productRefRepository.exists(8L)).isTrue();
+}
+```
+
+Note that the test uses a mock connection factory to avoid connecting to a running message broker.
+It also uses the Spring Amqp `MessageListener` to emit the event. Thus we leave the deserialization to Spring Amqp and thus also test the configuration there.
+See the [full test source](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/test/java/com/epages/checkout/ProductSubscriberHardcodedPayloadTest.java) for on how this is done in detail.
+
+Our basic test works for a start but as our service landscape becomes more and more complex, we will find that such an approach just does not scale.
+Eventually, we will make changes to the product service without anticipating that these changes will break the checkout service.
+The tests of both services will stay green although the interaction is broken.
+
+## Spring Cloud Contract and Provider Contracts
+
+Our situation would improve significantly if we could just test the event subscriber against the latest version of the payload as generated by the product service.
+At the same time, we still want to be able to run this test without having to start a message broker and the product service.
+
+Spring Cloud Contract is the missing piece here.
+It is an umbrella project to help users to implement the [consumer-driven contract approach](http://martinfowler.com/articles/consumerDrivenContracts.html).
+It addresses REST and messaging interactions.
+We will use its messaging functionalities for our sample application.
+Spring Cloud Contract is providing integrations with Spring Integration, Spring Cloud Stream, Apache Camel, and plain [Spring Amqp](https://cloud.spring.io/spring-cloud-contract/spring-cloud-contract.html#_stub_runner_spring_amqp).
+
+Testing the interaction between our checkout and product service with Spring Cloud Contract works as follows.
+
+Our product service expresses the provider contract in the contract definition language which is a groovy DSL that describes the event. This contract describes the structure of the event.
```

Comment:
This contract describes the structure of the event. > Here's the structure of the event: (to avoid double "describes")